librz/il/vm_layer.c

```diff
@@ -0,0 +1,484 @@
+			rz_il_free_variable(var);
+			vm->vm_global_variable_list[i] = NULL;
+		}
+	}
+	free(vm->vm_global_variable_list);
+
+	if (vm->ct_opcodes) {
+		ht_pp_free(vm->ct_opcodes);
+	}
+
+	if (vm->mems) {
+		for (int i = 0; i < vm->mem_count; ++i) {
+			rz_il_free_mem(vm->mems[i]);
+		}
+		free(vm->mems);
+	}
+
+	if (vm->temp_value_list != NULL) {
+		for (int i = 0; i < VM_MAX_TEMP; ++i) {
+			free(vm->temp_value_list[i]);
+		}
+		free(vm->temp_value_list);
+	}
+
+	if (vm->vm_global_bind_table != NULL) {
+		ht_pp_free(vm->vm_global_bind_table);
+	}
+
+	if (vm->vm_global_label_table != NULL) {
+		ht_pp_free(vm->vm_global_label_table);
+	}
+
+	if (vm->op_handler_table != NULL) {
+		free(vm->op_handler_table);
+	}
+	rz_il_bv_free(vm->pc);
+	free(vm);
+}
+
+// Step on core theory opcode
+void rz_il_vm_step(RzILVM vm, RzILOp op) {
+	vm->easy_debug += 1;
+	RzILOpHandler handler = vm->op_handler_table[op->code];
+	handler(vm, op);
+}
+
+/**
+ * Step on the instruction of the original arch
+ * the instruction will be translated to a vector of core theory ops
+ * \param vm RzILVM, pointer to VM
+ * \param op_list RzPVector*, pointer to the vector of core theory ops
+ */
+RZ_API void rz_il_vm_list_step(RzILVM vm, RzPVector *op_list) {
+	void **iter;
+	RzILOp cur_op;
+
+	// rz_il_vm_debug_print_ops(vm);
+	// rz_il_print_vm_labels(vm);
+
+	rz_pvector_foreach (op_list, iter) {
+		cur_op = *iter;
+		rz_il_vm_step(vm, cur_op);
+		//		printf(">>>> %s\n", rz_il_op2str(cur_op->code));
+		//		rz_il_print_vm_temps(vm);
+	}
+
+	BitVector one = rz_il_bv_new(vm->pc->len);
+	rz_il_bv_set(one, vm->pc->len - 1, true);
+	BitVector next_pc = rz_il_bv_add(vm->pc, one);
+	rz_il_bv_free(vm->pc);
+	rz_il_bv_free(one);
+	vm->pc = next_pc;
+}
+
+/**
+ * Convert to bitvector from ut64
+ * similar API in librz/il/definition/bitvector.h
+ * \param addr ut64, an address
+ * \return BitVector, 64-bit bitvector
+ */
+RZ_API BitVector rz_il_ut64_addr_to_bv(ut64 addr) {
+	return rz_il_bv_new_from_ut64(64, addr);
+}
+
+/**
+ * Convert to ut64 from bitvector
+ * similar API in librz/il/definition/bitvector.h
+ * \param addr BitVector, a bitvector address
+ * \return ut64, the value of bitvector
+ */
+RZ_API ut64 rz_il_bv_addr_to_ut64(BitVector addr) {
+	return rz_il_bv_to_ut64(addr);
+}
+
+/**
+ * the same as rz_il_bv_free, free a bitvector address
+ * \param addr BitVector, a bitvector to free
+ */
+RZ_API void rz_il_free_bv_addr(BitVector addr) {
+	rz_il_bv_free(addr);
+}
+
+/**
+ * Add a memory in VM. We design this to support multiple memory in the future
+ * \param vm RzILVM, pointer to VM
+ * \param min_unit_size int, size of minimal unit of the vm
+ * \return Mem memory, return a pointer to the newly created memory
+ */
+RZ_API Mem rz_il_vm_add_mem(RzILVM vm, int min_unit_size) {
+	Mem mem = rz_il_new_mem(min_unit_size);
+	vm->mems[vm->mem_count] = mem;
+	vm->mem_count += 1;
+	return mem;
+}
+
+RZ_API string rz_il_op2str(CoreTheoryOPCode opcode) {
+	char *ctops[64] = {
+		"VAR",
```

Comment:
Adding new items to `CoreTheoryOPCode` enum will break ouput of this function. You should use a better way to map values or at least add reminder to `CoreTheoryOPCode`.

Reply:
@Heersin don't forget about this please, you can make a an array with enum values as the index